File: ShapeByte/framework/device/device_size/DeviceSizeCategory.swift
//
//  DeviceSizeCategory.swift
//  ShapeByte
//
//  Created by Lang, Stefan [RTL Tech] on 13.08.24.
//

import UIKit

enum DeviceSizeCategory {
    case xSmall(size: CGSize)
    case small(size: CGSize)
    case medium(size: CGSize)
    case large(size: CGSize)

    var size: CGSize {
        let retVal: CGSize

        switch self {
        case .xSmall(let size):
            retVal = size
        case .small(let size):
            retVal = size
        case .medium(let size):
            retVal = size
        case .large(let size):
            retVal = size
        }

        return retVal
    }

    var isXSmall: Bool {
        let retVal: Bool

        switch self {
        case .xSmall:
            retVal = true
        default:
            retVal = false
        }

        return retVal
    }

    var isSmall: Bool {
        let retVal: Bool

        switch self {
        case .small:
            retVal = true
        default:
            retVal = false
        }

        return retVal
    }

    var isMedium: Bool {
        let retVal: Bool

        switch self {
        case .medium:
            retVal = true
        default:
            retVal = false
        }

        return retVal
    }

    var isLarge: Bool {
        let retVal: Bool

        switch self {
        case .large:
            retVal = true
        default:
            retVal = false
        }

        return retVal
    }
}

File: foundation/core/api/src/commonMain/kotlin/de/stefan/lang/foundationCore/api/devicesize/DeviceSizeCategory.kt
package de.stefan.lang.foundationCore.api.devicesize

sealed interface DeviceSizeCategory {
    val size: Size

    data object None : DeviceSizeCategory {
        override val size: Size
            get() = Size.ZERO
    }

    data class XSmall(override val size: Size) : DeviceSizeCategory
    data class Small(override val size: Size) : DeviceSizeCategory
    data class Medium(override val size: Size) : DeviceSizeCategory
    data class Large(override val size: Size) : DeviceSizeCategory

    val isXSmall: Boolean get() = this is XSmall
    val isSmall: Boolean get() = this is Small
    val isMedium: Boolean get() = this is Medium
    val isLarge: Boolean get() = this is Large
}
